File: Day14/part_1_2.js
const knotHash = require("../Day10/part_1_2");

const start = `flqrgnkx`;

let totalCount = 0;

function hashToBinary(hash) {
  const line = hash
    .split("")
    .map(c =>
      parseInt(c, 16)
        .toString(2)
        .padStart(4, "0"),
    )
    .join("");

  return line;
}

const map = {};

for (let i = 0; i < 128; i++) {
  const input = `${start}-${i}`;

  const line = hashToBinary(knotHash(input));

  if (line.length !== 128) {
    throw new Error("WTF");
  }

  const count = line.split("").reduce((acc, c) => {
    return acc + (c === "1" ? 1 : 0);
  }, 0);

  line
    .split("")
    .map(c => {
      if (c === "1") {
        return "#";
      }
      return ".";
    })
    .forEach((c, j) => {
      map[`x${j}y${i}`] = c;
    });

  totalCount += count;
}

let currentZone = 1;

function go(x, y, zone = -1) {
  const coord = `x${x}y${y}`;

  if (typeof map[coord] === "undefined") {
    return;
  }

  if (map[coord] === ".") {
    return;
  }

  if (typeof map[coord] === "number") {
    return;
  }

  if (map[coord] === "#") {
    if (zone === -1) {
      map[coord] = currentZone;
      currentZone += 1;
    }
    go(x - 1, y, currentZone);
    go(x + 1, y, currentZone);
    go(x, y - 1, currentZone);
    go(x, y + 1, currentZone);
  }
}

for (let x = 0; x < 128; x++) {
  for (let y = 0; y < 128; y++) {
    go(x, y);
  }
}

// go(0, 0);

console.log(currentZone);

// console.log(
//   "10100000110000100000000101110000".startsWith(hashToBinary("a0c2017")),
// );

// console.log(totalCount);

File: Day14/part_1_2.go
package Day14

import (
	"fmt"
	"strconv"

	"github.com/agentcooper/advent-of-code-2017/Day10"
)

// Part is either Part1 or Part2
type Part int

const (
	// Part1 is part 1
	Part1 Part = 1
	// Part2 is part 2
	Part2 Part = 2
)

func HashToBinary(hash string) string {
	out := ""
	for _, r := range hash {
		d, _ := strconv.ParseInt(string(r), 16, 64)
		out += fmt.Sprintf("%04b", d)
	}
	return out
}

type Coord struct {
	x int
	y int
}

type Cell struct {
	zone int
}

type Map map[Coord]*Cell

func (m Map) ClaimZone(coord Coord, zone int, currentZone *int) {
	if _, ok := m[coord]; !ok {
		return
	}

	if m[coord].zone > 0 {
		return
	}

	z := zone
	if zone == -1 {
		*currentZone++
		z = *currentZone
	}
	m[coord].zone = z

	m.ClaimZone(Coord{x: coord.x - 1, y: coord.y}, z, currentZone)
	m.ClaimZone(Coord{x: coord.x + 1, y: coord.y}, z, currentZone)
	m.ClaimZone(Coord{x: coord.x, y: coord.y - 1}, z, currentZone)
	m.ClaimZone(Coord{x: coord.x, y: coord.y + 1}, z, currentZone)
}

// Solve solves the puzzle
func Solve(input string, part Part) int {
	m := Map{}
	totalCount := 0

	for i := 0; i < 128; i++ {
		s := fmt.Sprintf("%s-%d", input, i)
		line := HashToBinary(Day10.KnotHash(s))

		for j, r := range line {
			if r == '1' {
				totalCount++
				m[Coord{x: i, y: j}] = &Cell{}
			}
		}
	}

	if part == Part1 {
		return totalCount
	}

	if part == Part2 {
		currentZone := 0

		for x := 0; x < 128; x++ {
			for y := 0; y < 128; y++ {
				m.ClaimZone(Coord{x: x, y: y}, -1, &currentZone)
			}
		}

		return currentZone
	}

	panic("Unknown part")
}
